fdbclient/AuditUtils.actor.cpp

```diff
@@ -109,31 +268,62 @@ ACTOR Future<UID> persistNewAuditState(Database cx,
 	ASSERT(!auditState.id.isValid());
 	state Transaction tr(cx);
 	state UID auditId;
-
-	loop {
-		try {
-			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
-			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-			wait(checkMoveKeysLock(&tr, lock, ddEnabled, true));
-			std::vector<AuditStorageState> auditStates = wait(getLatestAuditStatesImpl(&tr, auditState.getType(), 1));
-			uint64_t nextId = 1;
-			if (!auditStates.empty()) {
-				nextId = auditStates.front().id.first() + 1;
+	state AuditStorageState latestExistingAuditState;
+	TraceEvent(SevDebug, "AuditUtilPersistedNewAuditStateStart", auditId);
+	try {
+		loop {
+			try {
+				tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+				tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				wait(checkMoveKeysLock(&tr, lock, ddEnabled, true));
+				RangeResult res =
+				    wait(tr.getRange(auditKeyRange(auditState.getType()), 1, Snapshot::False, Reverse::True));
+				ASSERT(res.size() == 0 || res.size() == 1);
+				uint64_t nextId = 1;
+				if (res.size() == 1) {
+					latestExistingAuditState = decodeAuditStorageState(res[0].value);
+					if (auditId.isValid()) { // new audit state persist gets failed last time
+						// Check to confirm no other actor can persist new audit state
+						ASSERT(latestExistingAuditState.id.first() <= auditId.first());
```

Comment:
Is it possible that two requests come the same time, what will happen if the first one failed, could the second one call `persistNewAuditState()`?

Reply:
In this PR, if two requests come at the same time, the first one will proceed and the second one will be rejected by throwing `audit_storage_exceeded_request_limit`. If the first one is failed, the client will retry with a third request. The third request will be accepted.